CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+- Long term: Support a clustered, but un-sharded deployment mode with no cross-slot limitations, which will allow us to deprecate all Sentinel code
+- Long term: Enable centralized management for global configuration settings & Redis Functions
+
+This new approach is:
+- Based on a two-tier strongly consistent, consensus-based control plane
+- Maintains full backwards compatibility with data path Cluster commands (MOVED, NODES, SLOTS, ...), but not 
+  administrative commands (SETSLOT, FORGET, MEET,...)
+- Has a consistent source of truth for all cluster metadata
+- Requires the shards to reconcile their state according to the above source
+- Maintains the ability to scale down to a bare minimum 3 redis-server processes
+- Is called Flotilla :)
+
+And is intended to fully replace the current clustering implementation after a deprecation period in which the two 
+implementations will live side-by-side with compile time flag to chose between the two.
+
+## Motivation
+
+The current Redis Cluster is based on Gossip with the use of epochs to attempt to 
+control/reduce the impact of concurrent changes. It is easy to set up but has a lot of issues when performing day 2
+operations & scaling.
+
+For example:
+- Reliance on primaries (that may fail-over and move around) for quorum makes it impossible to guarantee availability using rack awareness.
+- Node removal has an inherent race condition which can lead to nodes being re-added
+- No single source of truth about who the primary is for a given slot range
+
+Our main goal with this new approach is to provide a more robust solution, with an easy-to-understand
+behavior and predictable failure modes. Beyond this, we also aim to substantially improve the scalability in order to
+be able to easily support a cluster with thousands of shards.
+
+## Design
+
+### Terminology
+
+| Term | Description |
+| --- | --- |
+| Node | A Single redis-server process |
+| Shard | A logical entity comprised of a group of nodes owning the same slots. It is automatically allocated when a new unique slot range is assigned to a node and automatically deallocated when all slots are moved away |
+| Host | A physical / virtual machine hosting the Flotilla processes |
+
+### High Level Overview
+
+Flotilla is composed of two tiers of strongly consistent, consensus-based control plane systems - The Topology Director 
+(TD) and the Failover Coordinator (FC).
+
+The Topology Director is the source of truth for shard existence and metadata, slot range allocation and Failover Coordinator
+assignments. In addition to the above, it also propagates failover information across the various Failover Coordinators.
+
+The Failover Coordinators is the source of truth for Primary/Replica roles. Each Failover Coordinators is assigned a subset 
+of the total nodes in the cluster, receives & propagates heartbeat information from them and enables them to make failover decisions in a
+consistent & timely manner.
```

Comment:
If by shard level data you mean things like owned slots, then not really.
We had this discussion when we talked about the possibility of rejecting invalid configs and said that while the TD is the source of truth for slot ownership, the FC implementation may reject a topology update which it perceives as invalid (for example, one which assigns an owned slot that still contains keys to another FC) as a failsafe in order to prevent bugs from causing data loss.

Reply:
I suppose let me phrase the context differently. I presume in the alternative consistency mode, you are achieving consensus on the stream of data being committed to the shard. Presumably the failover is then serializable within that stream of data. If the proposal is that the FC will act as an independent agent from that serializable sequence of updates, that doesn't make that much sense to me.

Also, agree with @PingXie about what happens in disagreements, it does need to defined (not now, but eventually). The current document is very handwavy. Saying, "we'll fix the bugs" is not a very strong defense for issues and how to recover from a wedged cluster.

Reply:
I think the general strategy here is actually pretty straight forward and is similar to what @PingXie wrote:
- What the FC owning a shard says about primary/replica status of the nodes belonging to that shard, wins.
- What the TD says about all other topology, wins.
(Since the TD propagates the information between the various FCs, there's no potential for multiple FCs to successfully publish status for the same shard -- and if/when we enable cross FC gossip as an enhancement it will still be possible to detect stale information based on the cluster epoch sent alongside it).

Way I see it, the handling of invalid config shouldn't be part of this spec -- and can be discussed in depth when we start working on the reference implementation.

Reply:
>Way I see it, the handling of invalid config shouldn't be part of this spec 

@ushachar agreed that the actual error handling can be part of the detailed design but the contract, such as how we ensure we will never get into a deadlock state should be part of the spec IMO. Otherwise, we might not have consistent experience across different implementations.

Reply:
The design ensures this by having the TD act as the single source of truth for all topology (other than primary/replica status). If an implementation allows the FC/Node to reject topology in order to prevent possible data loss due to TD side bugs it will need to specify how to fix that situation.

Not really a deadlock, but potentially relevant: Since the TD will block a small subset of changes to a particular shard until all the nodes in that shard have acked the current topology (see the reasoning under "Migrate a Shard between Failover Coordinators") a node/FC failure mid change would prevent subsequent topology changes in that shard until the node/FC comes back up. We will need to support a 'force' flag for admins when they know that the relevant node is dead and not coming back up.

Reply:
Added a placeholder for troubleshooting which mentions the need - will need to invest in this significantly when we start working on the actual implementation.